File: src/tools/featureFlags/create.ts
import { z } from "zod";
import type { Context, Tool } from "../types";
import { FilterGroupsSchema } from "../../schema/flags";
import { getProjectBaseUrl } from "../../lib/utils/api";

const schema = z.object({
	name: z.string(),
	key: z.string(),
	description: z.string(),
	filters: FilterGroupsSchema,
	active: z.boolean(),
	tags: z.array(z.string()).optional(),
});

type Params = z.infer<typeof schema>;

export const createHandler = async (context: Context, params: Params) => {
	const { name, key, description, filters, active, tags } = params;
	const projectId = await context.getProjectId();

	const flagResult = await context.api.featureFlags({ projectId }).create({
		data: { name, key, description, filters, active, tags },
	});

	if (!flagResult.success) {
		throw new Error(`Failed to create feature flag: ${flagResult.error.message}`);
	}

	const featureFlagWithUrl = {
		...flagResult.data,
		url: `${getProjectBaseUrl(projectId)}/feature_flags/${flagResult.data.id}`,
	};

	return {
		content: [{ type: "text", text: JSON.stringify(featureFlagWithUrl) }],
	};
};

const tool = (): Tool<typeof schema> => ({
	name: "create-feature-flag",
	description: `Creates a new feature flag in the project. Once you have created a feature flag, you should:
     - Ask the user if they want to add it to their codebase
     - Use the "search-docs" tool to find documentation on how to add feature flags to the codebase (search for the right language / framework)
     - Clarify where it should be added and then add it.
    `,
	schema,
	handler: createHandler,
});

export default tool;

File: python/tools/feature_flags/create.py
import json

from api.client import is_error, is_success
from lib.utils.api import get_project_base_url
from schema.tool_inputs import FeatureFlagCreateSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def create_feature_flag_handler(context: Context, params: FeatureFlagCreateSchema) -> ToolResult:
    project_id = await context.get_project_id()

    flag_result = await context.api.feature_flags(project_id).create(params)

    if is_error(flag_result):
        raise Exception(f"Failed to create feature flag: {flag_result.error}")

    assert is_success(flag_result)

    flag_data = flag_result.data

    feature_flag_with_url = {
        **flag_data.model_dump(),
        "url": f"{get_project_base_url(project_id, context.api.base_url)}/feature_flags/{flag_data.id}",
    }

    return ToolResult(content=json.dumps(feature_flag_with_url))


def create_feature_flag_tool() -> Tool[FeatureFlagCreateSchema]:
    definition = get_tool_definition("create-feature-flag")

    return Tool(
        name="create-feature-flag",
        description=definition["description"],
        schema=FeatureFlagCreateSchema,
        handler=create_feature_flag_handler,
    )
